### PR TITLE
Add support for closing of connections in the ABI/SDK.

### PR DESCRIPTION
**Pull request by jplevyak**
Signed-off-by: John Plevyak <jplevyak@gmail.com>

### CODE REVIEW

File: proxy_wasm_api.h [hunk 1]
```diff
@@ -591,8 +591,12 @@ inline WasmResult setFilterStateStringValue(StringView key, StringView s) {
 }
 
 // Continue/Respond/Route
-inline WasmResult continueRequest() { return proxy_continue_request(); }
-inline WasmResult continueResponse() { return proxy_continue_response(); }
+inline WasmResult continueRequest() { return proxy_continue_stream(WasmStreamType::Request); }
+inline WasmResult continueResponse() { return proxy_continue_stream(WasmStreamType::Response); }
+
+inline WasmResult closeRequest() { return proxy_close_stream(WasmStreamType::Request); }
+inline WasmResult closeResponse() { return proxy_continue_stream(WasmStreamType::Response); }
```
Comment (kyessenov) [inline proxy_wasm_api.h:598]: close_stream?

Comment (jplevyak) [reply proxy_wasm_api.h]: Thanx! fixed.

File: proxy_wasm_api.h [hunk 2]
```diff
@@ -591,8 +591,12 @@ inline WasmResult setFilterStateStringValue(StringView key, StringView s) {
 }
 
 // Continue/Respond/Route
-inline WasmResult continueRequest() { return proxy_continue_request(); }
-inline WasmResult continueResponse() { return proxy_continue_response(); }
+inline WasmResult continueRequest() { return proxy_continue_stream(WasmStreamType::Request); }
+inline WasmResult continueResponse() { return proxy_continue_stream(WasmStreamType::Response); }
+
+inline WasmResult closeRequest() { return proxy_close_stream(WasmStreamType::Request); }
+inline WasmResult closeResponse() { return proxy_close_stream(WasmStreamType::Response); }
```
Comment (PiotrSikora) [inline proxy_wasm_api.h:598]: I think `close` should be exposed for network filters only.

We generally don't expose underlying connection in HTTP (at least not yet), and there is no concept of closing request separately from closing response (unless we start doing retries within Proxy-Wasm, but that's not there yet). Also, we already have `send_local_response`, which effectively terminates HTTP request/response flow, and seems like much more elegant way to terminate HTTP request/response than hard close (e.g. there would be  response code and details in the logs).

Comment (jplevyak) [reply proxy_wasm_api.h:598]: What about the case where you are tunneling another protocol through say HTTP CONNECT or POST and there is a error in the inner protocol?  Do we want to jam an HTTP response in the middle of the request?  Is that legal for CONNECT?

Comment (PiotrSikora) [reply proxy_wasm_api.h:598]: Tunneling via HTTP/2 CONNECT is going to be handed off to normal network filters, not HTTP filters, so it won't be handled as HTTP request/response in Proxy-Wasm.

Comment (jplevyak) [reply proxy_wasm_api.h:598]: As discussed, I like the idea of being able to just shut something down, even if we are in the middle of say tunneling the response.

**Comment by jplevyak** [comment]
Ping.

**Pull request merged**